No parameter injector found for parameter 1 of route method

I'm not sure about this is a bug or not. Anyway, I'm getting the following error using Reactive Routes:

```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:1.9.2.Final:build (default) on project employee-quarkus: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.vertx.web.deployment.VertxWebProcessor#validateBeanDeployment threw an exception: java.lang.IllegalStateException: No parameter injector found for parameter 1 of route method void getEmployee(io.vertx.reactivex.ext.web.RoutingContext rc, java.lang.Long id) declared on CLASS bean [types=[org.serrodcal.poc.resource.EmployeeReactiveResource, java.lang.Object], qualifiers=[@Default, @Any], target=org.serrodcal.poc.resource.EmployeeReactiveResource]
[ERROR] 	at io.quarkus.vertx.web.deployment.VertxWebProcessor.validateRouteMethod(VertxWebProcessor.java:450)
[ERROR] 	at io.quarkus.vertx.web.deployment.VertxWebProcessor.validateBeanDeployment(VertxWebProcessor.java:166)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:936)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
[ERROR] 	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2046)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1578)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:834)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:479)
```

In the Resource, I annotated a method argument as `@Param` but Quarkus is not able to build the application:

```
@ApplicationScoped
@RouteBase(path = "api/v2")
public class EmployeeReactiveResource {

    @Inject
    EmployeeService employeeService;

   @Route(path = "employee/:id", methods = HttpMethod.GET)
    void getEmployee(RoutingContext rc, @Param("id") Long id) {
        //TODO
    }

}
```

Is it possible to use @Param to inject the value of the param?

I'm using Quarkus 1.9.2.Final and Java 11.

--- a/extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/Methods.java
+++ b/extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/Methods.java
@@ -1,5 +1,7 @@
 package io.quarkus.vertx.web.deployment;
 
+import java.util.Collection;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Optional;
 import java.util.Set;
@@ -196,6 +198,27 @@ class Methods {
     static final MethodDescriptor IS_ASSIGNABLE_FROM = MethodDescriptor.ofMethod(Class.class, "isAssignableFrom",
             boolean.class, Class.class);
     static final MethodDescriptor GET_CLASS = MethodDescriptor.ofMethod(Object.class, "getClass", Class.class);
+
+    static final MethodDescriptor STRING_CHAR_AT = MethodDescriptor.ofMethod(String.class, "charAt", int.class);
+    static final MethodDescriptor INTEGER_VALUE_OF = MethodDescriptor.ofMethod(Integer.class, "valueOf", Integer.class,
+            String.class);
+    static final MethodDescriptor LONG_VALUE_OF = MethodDescriptor.ofMethod(Long.class, "valueOf", Long.class, String.class);
+    static final MethodDescriptor BOOLEAN_VALUE_OF = MethodDescriptor.ofMethod(Boolean.class, "valueOf", Boolean.class,
+            String.class);
+    static final MethodDescriptor CHARACTER_VALUE_OF = MethodDescriptor.ofMethod(Character.class, "valueOf", Character.class,
+            char.class);
+    static final MethodDescriptor FLOAT_VALUE_OF = MethodDescriptor.ofMethod(Float.class, "valueOf", Float.class, String.class);
+    static final MethodDescriptor DOUBLE_VALUE_OF = MethodDescriptor.ofMethod(Double.class, "valueOf", Double.class,
+            String.class);
+    static final MethodDescriptor SHORT_VALUE_OF = MethodDescriptor.ofMethod(Short.class, "valueOf", Short.class, String.class);
+    static final MethodDescriptor BYTE_VALUE_OF = MethodDescriptor.ofMethod(Byte.class, "valueOf", Byte.class, String.class);
+
+    static final MethodDescriptor COLLECTION_SIZE = MethodDescriptor.ofMethod(Collection.class, "size", int.class);
+    static final MethodDescriptor COLLECTION_ITERATOR = MethodDescriptor.ofMethod(Collection.class, "iterator", Iterator.class);
+    static final MethodDescriptor COLLECTION_ADD = MethodDescriptor.ofMethod(Collection.class, "add", boolean.class,
+            Object.class);
+    static final MethodDescriptor ITERATOR_NEXT = MethodDescriptor.ofMethod(Iterator.class, "next", Object.class);
+    static final MethodDescriptor ITERATOR_HAS_NEXT = MethodDescriptor.ofMethod(Iterator.class, "hasNext", boolean.class);
 
     private Methods() {
         // Avoid direct instantiation

--- a/extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java
+++ b/extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java
@@ -80,6 +80,7 @@ import io.quarkus.gizmo.MethodCreator;
 import io.quarkus.gizmo.MethodDescriptor;
 import io.quarkus.gizmo.ResultHandle;
 import io.quarkus.gizmo.TryBlock;
+import io.quarkus.gizmo.WhileLoop;
 import io.quarkus.hibernate.validator.spi.BeanValidationAnnotationsBuildItem;
 import io.quarkus.runtime.LaunchMode;
 import io.quarkus.runtime.util.HashUtil;
@@ -91,7 +92,6 @@ import io.quarkus.vertx.http.deployment.VertxWebRouterBuildItem;
 import io.quarkus.vertx.http.deployment.devmode.NotFoundPageDisplayableEndpointBuildItem;
 import io.quarkus.vertx.http.deployment.devmode.RouteDescriptionBuildItem;
 import io.quarkus.vertx.http.runtime.HandlerType;
-import io.quarkus.vertx.web.Header;
 import io.quarkus.vertx.web.Param;
 import io.quarkus.vertx.web.Route;
 import io.quarkus.vertx.web.RouteFilter;
@@ -1202,90 +1202,22 @@ class VertxWebProcessor {
                             }
                         }).build());
 
-        injectors.add(ParameterInjector.builder().matchType(io.quarkus.arc.processor.DotNames.STRING)
-                .matchType(ParameterizedType.create(io.quarkus.arc.processor.DotNames.OPTIONAL,
-                        new Type[] { Type.create(io.quarkus.arc.processor.DotNames.STRING, Kind.CLASS) }, null))
-                .matchType(ParameterizedType.create(DotNames.LIST,
-                        new Type[] { Type.create(io.quarkus.arc.processor.DotNames.STRING, Kind.CLASS) }, null))
+        injectors.add(ParameterInjector.builder().matchPrimitiveWrappers()
+                .matchType(io.quarkus.arc.processor.DotNames.STRING)
+                .matchOptionalOf(io.quarkus.arc.processor.DotNames.STRING)
+                .matchListOf(io.quarkus.arc.processor.DotNames.STRING)
                 .requireAnnotations(PARAM)
-                .resultHandleProvider(new ResultHandleProvider() {
-                    @Override
-                    public ResultHandle get(MethodInfo method, Type paramType, Set<AnnotationInstance> annotations,
-                            ResultHandle routingContext, MethodCreator invoke, int position,
-                            BuildProducer<ReflectiveHierarchyBuildItem> reflectiveHierarchy) {
-                        AnnotationValue paramAnnotationValue = Annotations
-                                .find(annotations, PARAM).value();
-                        String paramName = paramAnnotationValue != null ? paramAnnotationValue.asString() : null;
-                        if (paramName == null || paramName.equals(Param.ELEMENT_NAME)) {
-                            paramName = method.parameterName(position);
-                        }
-                        if (paramName == null) {
-                            throw parameterNameNotAvailable(position, method);
-                        }
-                        ResultHandle paramHandle;
-                        if (paramType.name().equals(DotNames.LIST)) {
-                            // routingContext.request().params().getAll(paramName)
-                            paramHandle = invoke.invokeInterfaceMethod(Methods.MULTIMAP_GET_ALL,
-                                    invoke.invokeInterfaceMethod(Methods.REQUEST_PARAMS, invoke
-                                            .invokeInterfaceMethod(Methods.REQUEST,
-                                                    routingContext)),
-                                    invoke.load(paramName));
-                        } else {
-                            // routingContext.request().getParam(paramName)
-                            paramHandle = invoke.invokeInterfaceMethod(Methods.REQUEST_GET_PARAM, invoke
-                                    .invokeInterfaceMethod(Methods.REQUEST,
-                                            routingContext),
-                                    invoke.load(paramName));
-                            if (paramType.name().equals(io.quarkus.arc.processor.DotNames.OPTIONAL)) {
-                                paramHandle = invoke.invokeStaticMethod(Methods.OPTIONAL_OF_NULLABLE, paramHandle);
-                            }
-                        }
-                        return paramHandle;
-                    }
-                }).build());
+                .resultHandleProvider(new ParamAndHeaderProvider(PARAM, Methods.REQUEST_PARAMS, Methods.REQUEST_GET_PARAM))
+                .build());
 
-        injectors.add(ParameterInjector.builder().matchType(io.quarkus.arc.processor.DotNames.STRING)
-                .matchType(ParameterizedType.create(io.quarkus.arc.processor.DotNames.OPTIONAL,
-                        new Type[] { Type.create(io.quarkus.arc.processor.DotNames.STRING, Kind.CLASS) }, null))
-                .matchType(ParameterizedType.create(DotNames.LIST,
-                        new Type[] { Type.create(io.quarkus.arc.processor.DotNames.STRING, Kind.CLASS) }, null))
+        injectors.add(ParameterInjector.builder().matchPrimitiveWrappers()
+                .matchType(io.quarkus.arc.processor.DotNames.STRING)
+                .matchOptionalOf(io.quarkus.arc.processor.DotNames.STRING)
+                .matchListOf(io.quarkus.arc.processor.DotNames.STRING)
                 .requireAnnotations(DotNames.HEADER)
-                .resultHandleProvider(new ResultHandleProvider() {
-                    @Override
-                    public ResultHandle get(MethodInfo method, Type paramType, Set<AnnotationInstance> annotations,
-                            ResultHandle routingContext, MethodCreator invoke, int position,
-                            BuildProducer<ReflectiveHierarchyBuildItem> reflectiveHierarchy) {
-                        AnnotationValue paramAnnotationValue = Annotations
-                                .find(annotations, DotNames.HEADER).value();
-                        String paramName = paramAnnotationValue != null ? paramAnnotationValue.asString() : null;
-                        if (paramName == null || paramName.equals(Header.ELEMENT_NAME)) {
-                            paramName = method.parameterName(position);
-                        }
-                        if (paramName == null) {
-                            throw parameterNameNotAvailable(position, method);
-                        }
-                        ResultHandle paramHandle;
-                        if (paramType.name().equals(DotNames.LIST)) {
-                            // routingContext.request().headers().getAll(paramName)
-                            paramHandle = invoke.invokeInterfaceMethod(Methods.MULTIMAP_GET_ALL,
-                                    invoke.invokeInterfaceMethod(Methods.REQUEST_HEADERS, invoke
-                                            .invokeInterfaceMethod(Methods.REQUEST,
-                                                    routingContext)),
-                                    invoke.load(paramName));
-                        } else {
-                            // routingContext.request().getHeader(paramName)
-                            paramHandle = invoke.invokeInterfaceMethod(Methods.REQUEST_GET_HEADER, invoke
-                                    .invokeInterfaceMethod(Methods.REQUEST,
-                                            routingContext),
-                                    invoke.load(paramName));
-                            if (paramType.name().equals(io.quarkus.arc.processor.DotNames.OPTIONAL)) {
-                                paramHandle = invoke.invokeStaticMethod(Methods.OPTIONAL_OF_NULLABLE, paramHandle);
-                            }
-
-                        }
-                        return paramHandle;
-                    }
-                }).build());
+                .resultHandleProvider(
+                        new ParamAndHeaderProvider(DotNames.HEADER, Methods.REQUEST_HEADERS, Methods.REQUEST_GET_HEADER))
+                .build());
 
         injectors
                 .add(ParameterInjector.builder()
@@ -1367,6 +1299,87 @@ class VertxWebProcessor {
                 "Unable to determine the name of the parameter at position " + position + " in method "
                         + method.declaringClass().name() + "#" + method.name()
                         + "() - compile the class with debug info enabled (-g) or parameter names recorded (-parameters), or specify the appropriate annotation value");
+
+    }
+
+    private static class ParamAndHeaderProvider implements ResultHandleProvider {
+
+        private final DotName annotationName;
+        private final MethodDescriptor multiMapAccessor;
+        private final MethodDescriptor valueAccessor;
+
+        public ParamAndHeaderProvider(DotName annotationName, MethodDescriptor multiMapAccessor,
+                MethodDescriptor valueAccessor) {
+            this.annotationName = annotationName;
+            this.multiMapAccessor = multiMapAccessor;
+            this.valueAccessor = valueAccessor;
+        }
+
+        @Override
+        public ResultHandle get(MethodInfo method, Type paramType, Set<AnnotationInstance> annotations,
+                ResultHandle routingContext, MethodCreator invoke, int position,
+                BuildProducer<ReflectiveHierarchyBuildItem> reflectiveHierarchy) {
+            AnnotationValue paramAnnotationValue = Annotations
+                    .find(annotations, annotationName).value();
+            String paramName = paramAnnotationValue != null ? paramAnnotationValue.asString() : null;
+            if (paramName == null || paramName.equals(Param.ELEMENT_NAME)) {
+                paramName = method.parameterName(position);
+            }
+            if (paramName == null) {
+                throw parameterNameNotAvailable(position, method);
+            }
+            AssignableResultHandle paramHandle = invoke.createVariable(Object.class);
+            if (paramType.name().equals(DotNames.LIST)) {
+                Type wrappedType = paramType.asParameterizedType().arguments().get(0);
+                // List<String> params = routingContext.request().params().getAll(paramName)
+                invoke.assign(paramHandle, invoke.invokeInterfaceMethod(Methods.MULTIMAP_GET_ALL,
+                        invoke.invokeInterfaceMethod(multiMapAccessor, invoke
+                                .invokeInterfaceMethod(Methods.REQUEST,
+                                        routingContext)),
+                        invoke.load(paramName)));
+                if (!wrappedType.name()
+                        .equals(io.quarkus.arc.processor.DotNames.STRING)) {
+                    // Iterate over the list and convert wrapped values
+                    ResultHandle results = invoke.newInstance(
+                            MethodDescriptor.ofConstructor(ArrayList.class, int.class),
+                            invoke.invokeInterfaceMethod(Methods.COLLECTION_SIZE, paramHandle));
+                    ResultHandle iterator = invoke.invokeInterfaceMethod(Methods.COLLECTION_ITERATOR, paramHandle);
+                    WhileLoop loop = invoke.whileLoop(new Function<BytecodeCreator, BranchResult>() {
+                        @Override
+                        public BranchResult apply(BytecodeCreator bc) {
+                            return bc.ifTrue(bc.invokeInterfaceMethod(Methods.ITERATOR_HAS_NEXT, iterator));
+                        }
+                    });
+                    BytecodeCreator block = loop.block();
+                    AssignableResultHandle element = block.createVariable(Object.class);
+                    block.assign(element, block.invokeInterfaceMethod(Methods.ITERATOR_NEXT, iterator));
+                    convertPrimitiveAndSet(element, wrappedType, block, method, position);
+                    block.invokeInterfaceMethod(Methods.COLLECTION_ADD, results, element);
+                    invoke.assign(paramHandle, results);
+                }
+            } else {
+                // Object param = routingContext.request().getParam(paramName)
+                invoke.assign(paramHandle, invoke.invokeInterfaceMethod(valueAccessor, invoke
+                        .invokeInterfaceMethod(Methods.REQUEST,
+                                routingContext),
+                        invoke.load(paramName)));
+                if (paramType.name().equals(io.quarkus.arc.processor.DotNames.OPTIONAL)) {
+                    Type wrappedType = paramType.asParameterizedType().arguments().get(0);
+                    if (!wrappedType.name()
+                            .equals(io.quarkus.arc.processor.DotNames.STRING)) {
+                        convertPrimitiveAndSet(paramHandle, wrappedType, invoke, method,
+                                position);
+                    }
+                    invoke.assign(paramHandle,
+                            invoke.invokeStaticMethod(Methods.OPTIONAL_OF_NULLABLE, paramHandle));
+                } else {
+                    if (!paramType.name().equals(io.quarkus.arc.processor.DotNames.STRING)) {
+                        convertPrimitiveAndSet(paramHandle, paramType, invoke, method, position);
+                    }
+                }
+            }
+            return paramHandle;
+        }
 
     }
 
@@ -1481,6 +1494,38 @@ class VertxWebProcessor {
                 return this;
             }
 
+            Builder matchPrimitiveWrappers() {
+                List<DotName> primitiveNames = Arrays.asList(io.quarkus.arc.processor.DotNames.INTEGER,
+                        io.quarkus.arc.processor.DotNames.LONG, io.quarkus.arc.processor.DotNames.SHORT,
+                        io.quarkus.arc.processor.DotNames.BOOLEAN, io.quarkus.arc.processor.DotNames.BYTE,
+                        io.quarkus.arc.processor.DotNames.CHARACTER, io.quarkus.arc.processor.DotNames.DOUBLE,
+                        io.quarkus.arc.processor.DotNames.FLOAT);
+                for (DotName name : primitiveNames) {
+                    matchType(name);
+                    matchOptionalOf(name);
+                    matchListOf(name);
+                }
+                return this;
+            }
+
+            Builder matchOptionalOf(DotName className) {
+                return matchOptionalOf(Type.create(className, Kind.CLASS));
+            }
+
+            Builder matchListOf(DotName className) {
+                return matchListOf(Type.create(className, Kind.CLASS));
+            }
+
+            Builder matchOptionalOf(Type type) {
+                return matchType(ParameterizedType.create(io.quarkus.arc.processor.DotNames.OPTIONAL,
+                        new Type[] { type }, null));
+            }
+
+            Builder matchListOf(Type type) {
+                return matchType(ParameterizedType.create(DotNames.LIST,
+                        new Type[] { type }, null));
+            }
+
             Builder skipType(DotName className) {
                 return skipType(Type.create(className, Kind.CLASS));
             }
@@ -1519,6 +1564,53 @@ class VertxWebProcessor {
 
         }
 
+    }
+
+    static void convertPrimitiveAndSet(AssignableResultHandle paramHandle, Type paramType, BytecodeCreator invoke,
+            MethodInfo method, int position) {
+        // For example: 
+        // if(param != null) { 
+        //    try {
+        //       param = Long.valueOf(param);
+        //    } catch(Throwable e) {
+        //       ...
+        //    }
+        // }
+        BytecodeCreator notNull = invoke.ifNotNull(paramHandle).trueBranch();
+        TryBlock tryBlock = notNull.tryBlock();
+        CatchBlockCreator catchBlock = tryBlock.addCatch(Throwable.class);
+        catchBlock.throwException(IllegalArgumentException.class,
+                "Error converting parameter at position " + position + " of method " + method,
+                catchBlock.getCaughtException());
+
+        if (paramType.name().equals(io.quarkus.arc.processor.DotNames.INTEGER)) {
+            tryBlock.assign(paramHandle, tryBlock.invokeStaticMethod(Methods.INTEGER_VALUE_OF,
+                    paramHandle));
+        } else if (paramType.name().equals(io.quarkus.arc.processor.DotNames.LONG)) {
+            tryBlock.assign(paramHandle, tryBlock.invokeStaticMethod(Methods.LONG_VALUE_OF,
+                    paramHandle));
+        } else if (paramType.name().equals(io.quarkus.arc.processor.DotNames.BOOLEAN)) {
+            tryBlock.assign(paramHandle, tryBlock.invokeStaticMethod(Methods.BOOLEAN_VALUE_OF,
+                    paramHandle));
+        } else if (paramType.name().equals(io.quarkus.arc.processor.DotNames.CHARACTER)) {
+            ResultHandle firstChar = tryBlock.invokeVirtualMethod(Methods.STRING_CHAR_AT, paramHandle, tryBlock.load(0));
+            tryBlock.assign(paramHandle, tryBlock.invokeStaticMethod(Methods.CHARACTER_VALUE_OF,
+                    firstChar));
+        } else if (paramType.name().equals(io.quarkus.arc.processor.DotNames.FLOAT)) {
+            tryBlock.assign(paramHandle, tryBlock.invokeStaticMethod(Methods.FLOAT_VALUE_OF,
+                    paramHandle));
+        } else if (paramType.name().equals(io.quarkus.arc.processor.DotNames.DOUBLE)) {
+            tryBlock.assign(paramHandle, tryBlock.invokeStaticMethod(Methods.DOUBLE_VALUE_OF,
+                    paramHandle));
+        } else if (paramType.name().equals(io.quarkus.arc.processor.DotNames.BYTE)) {
+            tryBlock.assign(paramHandle, tryBlock.invokeStaticMethod(Methods.BYTE_VALUE_OF,
+                    paramHandle));
+        } else if (paramType.name().equals(io.quarkus.arc.processor.DotNames.SHORT)) {
+            tryBlock.assign(paramHandle, tryBlock.invokeStaticMethod(Methods.SHORT_VALUE_OF,
+                    paramHandle));
+        } else {
+            throw new IllegalArgumentException("Unsupported param type: " + paramType);
+        }
     }
 
     interface ResultHandleProvider {

--- a/extensions/vertx-web/deployment/src/test/java/io/quarkus/vertx/web/params/RouteMethodParametersTest.java
+++ b/extensions/vertx-web/deployment/src/test/java/io/quarkus/vertx/web/params/RouteMethodParametersTest.java
@@ -48,7 +48,8 @@ public class RouteMethodParametersTest {
                 .body(is("Hello 42! Your name is foo and id is 10"));
         when().get("/hello-multiple-params?id=foo&id=10").then().statusCode(200)
                 .body(is("foo,10"));
-        given().header("My-Header", "fooooo").get("/hello-header").then().statusCode(200).body(is("fooooo"));
+        given().header("My-Header", "fooooo").header("My-Number-Header", "10000").get("/hello-header").then().statusCode(200)
+                .body(is("fooooo=10000"));
         given().header("My-Header", "fooooo").header("My-Header", "baaar").get("/hello-multiple-headers").then().statusCode(200)
                 .body(is("fooooo,baaar"));
         given().contentType("application/json").body("{\"name\":\"Eleven\"}")
@@ -67,6 +68,12 @@ public class RouteMethodParametersTest {
         given().contentType("application/json").body("{\"name\":\"Eleven\"}")
                 .post("/hello-body-pojo?id=13").then().statusCode(200).body("name", is("Eleven"))
                 .body("id", is(13));
+        when().get("/hello-params-conversion?id=22&size=100&valid=false&doubles=2").then().statusCode(200)
+                .body(is("id=22,size=100,valid=false,doubles=[2.0]"));
+        when().get("/hello-param-conversion-optional").then().statusCode(200)
+                .body(is("hello 42"));
+        when().get("/hello-param-conversion-optional?id=1").then().statusCode(200)
+                .body(is("hello 1"));
     }
 
     static class SimpleBean {
@@ -121,9 +128,10 @@ public class RouteMethodParametersTest {
         }
 
         @Route
-        String helloHeader(@Header("My-Header") String myHeader, @Header Optional<String> missingHeader) {
+        String helloHeader(@Header("My-Header") String myHeader, @Header Optional<String> missingHeader,
+                @Header("My-Number-Header") Long numberHeader) {
             assertFalse(missingHeader.isPresent());
-            return myHeader;
+            return myHeader + "=" + numberHeader;
         }
 
         @Route
@@ -155,6 +163,21 @@ public class RouteMethodParametersTest {
         Person helloBodyPojo(@Body Person person, @Param("id") Optional<String> primaryKey) {
             person.setId(primaryKey.map(Integer::valueOf).orElse(11));
             return person;
+        }
+
+        @Route
+        String helloParamsConversion(@Param Integer id, @Param Long size, @Param Boolean valid, @Param List<Double> doubles) {
+            return String.format("id=%s,size=%s,valid=%s,doubles=%s", id, size, valid, doubles);
+        }
+
+        @Route
+        String helloParamConversionOptional(@Param Optional<Integer> id) {
+            return "hello " + id.orElse(42);
+        }
+
+        @Route
+        String helloParamConversionFailure(@Param Integer id) {
+            return "hello " + id;
         }
 
     }

